### PR TITLE
Copy LruMap from scala-maxmind-iplookups

### PR DESCRIPTION
**Pull request by three**

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ */
+lazy val root = project
+  .in(file("."))
+  .settings(
+    organization := "com.snowplowanalytics",
+    name := "lru-map",
+    version := "0.1.0",
+    description := "Simple LRU Map for caching",
+    scalaVersion := "2.11.12",
+    crossScalaVersions := Seq("2.11.11", "2.12.2"),
```
Comment (BenFradet) [inline build.sbt]: 2.12.6

File: src/main/scala/com.snowplowanalytics.maxmind.lrumap/LruMap.scala [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+import java.util.{LinkedHashMap, Map}
+import scala.collection.JavaConverters._
+
+// Based on com.twitter.util.LruMap
+// https://github.com/twitter/util/blob/develop/util-collection/src/main/scala/com/twitter/util/LruMap.scala
+
+object LruMap {
+  def create[F[_]: Sync, K, V](size: Int): F[LruMap[F, K, V]] =
+    Sync[F].flatMap(makeUnderlying[F, K, V](size))(underlying =>
+      Sync[F].delay {
```
Comment (BenFradet) [inline src/main/scala/com.snowplowanalytics.maxmind.lrumap/LruMap.scala]: I don't think this is needed either?

File: README.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        } yield m1+m2
+      })
+  }
+}
+
+val result = (for {
+  // When the size of the map exceeds 500 the least recently used element is
+  // removed
+  lru <- LruMap.create[IO, BigInt, BigInt](500)
+  fib <- IO { new Fibonacci(lru) }
```
Comment (BenFradet) [inline README.md]: this is just memory allocation, I don't think we need to consider that a side-effect

File: src/main/scala/com.snowplowanalytics.maxmind.lrumap/LruMap.scala [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+package com.snowplowanalytics.lrumap
+
+import cats.effect.Sync
+import java.util.{LinkedHashMap, Map}
+import scala.collection.JavaConverters._
+
+// Based on com.twitter.util.LruMap
+// https://github.com/twitter/util/blob/develop/util-collection/src/main/scala/com/twitter/util/LruMap.scala
+
+object LruMap {
```
Comment (BenFradet) [inline src/main/scala/com.snowplowanalytics.maxmind.lrumap/LruMap.scala]: can you add some scaladoc to the object / class and functions please?

File: src/main/scala/com.snowplowanalytics.maxmind.lrumap/LruMap.scala [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+  // initial capacity and load factor are the normal defaults for LinkedHashMap
+  private def makeUnderlying[F[_]: Sync, K, V](maxSize: Int): F[ImpureLruMap[K, V]] = Sync[F].delay {
+    new ImpureLruMap[K, V](maxSize, 16, 0.75f)
+  }
+}
+
+/**
+ * A scala `Map` backed by a [[java.util.LinkedHashMap]]
+ */
+class LruMap[F[_]: Sync, K, V] private (val maxSize: Int, val underlying: ImpureLruMap[K, V]) {
```
Comment (BenFradet) [inline src/main/scala/com.snowplowanalytics.maxmind.lrumap/LruMap.scala]: why is `maxSize` needed? also no need to have `val`s

**Review by BenFradet — APPROVED** [review]
mostly nits and docs :+1:

**Comment by BenFradet** [comment]
LGTM, I'll let @chuwy have a second look

File: LICENSE-2.0.txt [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+      To apply the Apache License to your work, attach the following
+      boilerplate notice, with the fields enclosed by brackets "[]"
+      replaced with your own identifying information. (Don't include
+      the brackets!)  The text should be enclosed in the appropriate
+      comment syntax for the file format. We also recommend that a
+      file or class name and description of purpose be included on the
+      same "printed page" as the copyright notice for easier
+      identification within third-party archives.
+
+   Copyright 2012 Snowplow Analytics Ltd.
```
Comment (chuwy) [inline LICENSE-2.0.txt]: 2012-2018?

File: README.md [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+      })
+  }
+}
+
+val result = (for {
+  // When the size of the map exceeds 500 the least recently used element is
+  // removed
+  lru    <- LruMap.create[IO, BigInt, BigInt](500)
+  result <- (new Fibonacci(lru)).calc(100)
+} yield result).unsafeRunSync
```
Comment (chuwy) [inline README.md]: It is a very common convention to append `()` to `Unit`-side-effecting method invocation: `effect.unsafeRunSync()`

**Review by chuwy — APPROVED** [review]
Looks perfect.

**Comment by BenFradet** [comment]
Feel free to squash everything and tag a 0.1.0-M1

**Comment by BenFradet** [comment]
Please don't merge yourself there was still a lot that needed doing before going to master

**Comment by three** [comment]
Sorry! I thought you meant you wanted me to squash into master. Should I revert master back to 177f58f?

**Comment by BenFradet** [comment]
no I just meant squash here, no need to revert, please add a CHANGELOG and rename the commit to "Initial commit".

**Comment by BenFradet** [comment]
Also we usually don't merge before we have the go ahead to publish a final version e.g. not an release candidate or milestone version.

**Comment by three** [comment]
I squashed everything to one commit with the changelog and included it in the initial branch. Is there anything else I should do to get the final version ready?

**Comment by BenFradet** [comment]
This should be on master since the initial branch has already been merged...

**Comment by three** [comment]
Okay, I pushed the commit to master.

**Comment by BenFradet** [comment]
heads up that the changelog should look like

```
Version 0.1.0 (2017-08-31)
--------------------------
Initial release
```

but we need to wait on #2 and #3 to release a 0.1.0-M1 and then a final version.

**Comment by three** [comment]
I edited the changelog and put the squashed commit on master.

**Pull request merged**